### PR TITLE
Add tests for talker/listener examples

### PR DESCRIPTION
**Pull request by gerkey**
CI jobs forthcoming

Connects to ros2/ros2#143

Depends on https://github.com/ros2/launch/pull/12

Depends on https://github.com/ros2/launch/pull/13

### CODE REVIEW

**Comment by gerkey** [comment]
[Windows](http://ci.ros2.org/job/ros2_batch_ci_windows/638/console) times out:

~~~
01:58:31 test 7
01:58:31       Start  7: test_example_talker_listener__rmw_opensplice_cpp
01:58:31 
01:58:31 7: Test command: C:\tools\python\python.exe "-u" "C:/Jenkins/workspace/ros2_batch_ci_windows/workspace/install/ament_cmake_test/share/ament_cmake_test/cmake/run_test.py" "C:/Jenkins/workspace/ros2_batch_ci_windows/workspace/build/rclcpp_examples/test_results/rclcpp_examples/test_example_talker_listener__rmw_opensplice_cpp.xunit.xml" "--output-file" "C:/Jenkins/workspace/ros2_batch_ci_windows/workspace/build/rclcpp_examples/ament_cmake_nose/test_example_talker_listener__rmw_opensplice_cpp.txt" "--command" "C:/tools/python/Scripts/nosetests-3.4.exe" "C:/Jenkins/workspace/ros2_batch_ci_windows/workspace/build/rclcpp_examples/test_talker_listener__rmw_opensplice_cpp.py" "--with-xunit" "--xunit-file=C:/Jenkins/workspace/ros2_batch_ci_windows/workspace/build/rclcpp_examples/test_results/rclcpp_examples/test_example_talker_listener__rmw_opensplice_cpp.xunit.xml" "--xunit-testsuite-name=rclcpp_examples.nosetests"
01:58:31 7: Test timeout computed to be: 30
01:58:31 7: -- run_test.py: invoking following command in 'C:\Jenkins\workspace\ros2_batch_ci_windows\workspace\src\ros2\examples\rclcpp_examples':
01:58:31 7:  - C:/tools/python/Scripts/nosetests-3.4.exe C:/Jenkins/workspace/ros2_batch_ci_windows/workspace/build/rclcpp_examples/test_talker_listener__rmw_opensplice_cpp.py --with-xunit --xunit-file=C:/Jenkins/workspace/ros2_batch_ci_windows/workspace/build/rclcpp_examples/test_results/rclcpp_examples/test_example_talker_listener__rmw_opensplice_cpp.xunit.xml --xunit-testsuite-name=rclcpp_examples.nosetests
01:59:01  7/12 Test  #7: test_example_talker_listener__rmw_opensplice_cpp ................***Timeout  30.01 sec
~~~

[OSX](http://ci.ros2.org/job/ros2_batch_ci_osx/472/console) fails with mis-matched console output:

~~~
01:58:37 test 19
01:58:37       Start 19: test_example_talker_listener__rmw_connext_cpp
01:58:37 
01:58:37 19: Test command: /Users/osrf/jenkins/workspace/ros2_batch_ci_osx/workspace/venv/bin/python3 "-u" "/Users/osrf/jenkins/workspace/ros2_batch_ci_osx/workspace/install/ament_cmake_test/share/ament_cmake_test/cmake/run_test.py" "/Users/osrf/jenkins/workspace/ros2_batch_ci_osx/workspace/build/rclcpp_examples/test_results/rclcpp_examples/test_example_talker_listener__rmw_connext_cpp.xunit.xml" "--output-file" "/Users/osrf/jenkins/workspace/ros2_batch_ci_osx/workspace/build/rclcpp_examples/ament_cmake_nose/test_example_talker_listener__rmw_connext_cpp.txt" "--command" "/Users/osrf/jenkins/workspace/ros2_batch_ci_osx/workspace/venv/bin/nosetests-3.4" "/Users/osrf/jenkins/workspace/ros2_batch_ci_osx/workspace/build/rclcpp_examples/test_talker_listener__rmw_connext_cpp.py" "--with-xunit" "--xunit-file=/Users/osrf/jenkins/workspace/ros2_batch_ci_osx/workspace/build/rclcpp_examples/test_results/rclcpp_examples/test_example_talker_listener__rmw_connext_cpp.xunit.xml" "--xunit-testsuite-name=rclcpp_examples.nosetests"
01:58:37 19: Test timeout computed to be: 30
01:58:38 19: -- run_test.py: invoking following command in '/Users/osrf/jenkins/workspace/ros2_batch_ci_osx/workspace/src/ros2/examples/rclcpp_examples':
01:58:38 19:  - /Users/osrf/jenkins/workspace/ros2_batch_ci_osx/workspace/venv/bin/nosetests-3.4 /Users/osrf/jenkins/workspace/ros2_batch_ci_osx/workspace/build/rclcpp_examples/test_talker_listener__rmw_connext_cpp.py --with-xunit --xunit-file=/Users/osrf/jenkins/workspace/ros2_batch_ci_osx/workspace/build/rclcpp_examples/test_results/rclcpp_examples/test_example_talker_listener__rmw_connext_cpp.xunit.xml --xunit-testsuite-name=rclcpp_examples.nosetests
01:58:45 19: F
01:58:45 19: ======================================================================
01:58:45 19: FAIL: test_talker_listener__rmw_connext_cpp.test_executable
01:58:45 19: ----------------------------------------------------------------------
01:58:45 19: Traceback (most recent call last):
01:58:45 19:   File "/Users/osrf/jenkins/workspace/ros2_batch_ci_osx/workspace/venv/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
01:58:45 19:     self.test(*self.arg)
01:58:45 19:   File "/Users/osrf/jenkins/workspace/ros2_batch_ci_osx/workspace/build/rclcpp_examples/test_talker_listener__rmw_connext_cpp.py", line 96, in test_executable
01:58:45 19:     (output_lines, expected_output_lines)
01:58:45 19: nose.proxy.AssertionError: Example output ([b'I heard: [Hello World: 2]', b'signal SIGINT']) does not match expected output ([b'I heard: [Hello World: 1]', b'I heard: [Hello World: 2]'])
01:58:45 19: -------------------- >> begin captured stdout << ---------------------
01:58:45 19: (test_executable_0) pid 36185: ['/Users/osrf/jenkins/workspace/ros2_batch_ci_osx/workspace/build/rclcpp_examples/talker__rmw_connext_cpp', 'test_executable'] (InMemoryHandler)
01:58:45 19: (test_executable_1) pid 36186: ['/Users/osrf/jenkins/workspace/ros2_batch_ci_osx/workspace/build/rclcpp_examples/listener__rmw_connext_cpp', 'test_executable'] (InMemoryHandler)
01:58:45 19: (test_executable_0) rc 0
01:58:45 19: () tear down
01:58:45 19: (test_executable_1) signal SIGINT
01:58:45 19: (test_executable_1) rc 0
01:58:45 19: 
01:58:45 19: --------------------- >> end captured stdout << ----------------------
01:58:45 19: -------------------- >> begin captured logging << --------------------
01:58:45 19: asyncio: DEBUG: Using selector: KqueueSelector
01:58:45 19: --------------------- >> end captured logging << ---------------------
01:58:45 19: 
01:58:45 19: ----------------------------------------------------------------------
01:58:45 19: Ran 1 test in 7.113s
01:58:45 19: 
01:58:45 19: FAILED (failures=1)
01:58:45 19: -- run_test.py: return code 1
01:58:45 19: -- run_test.py: verify result file '/Users/osrf/jenkins/workspace/ros2_batch_ci_osx/workspace/build/rclcpp_examples/test_results/rclcpp_examples/test_example_talker_listener__rmw_connext_cpp.xunit.xml'
01:58:45 19/26 Test #19: test_example_talker_listener__rmw_connext_cpp ........................***Failed    7.56 sec
~~~

**Comment by gerkey** [comment]
After 2c7ec2c, I expect the talker/listener test to work on OpenSplice (including Windows), but to fail with Connext (likely all 3 platforms).  Still need to resolve how to handle variable output (when listener misses the very first "hello world" message).

**Comment by gerkey** [comment]
Indeed, [Linux](ci.ros2.org/job/ros2_batch_ci_linux/588/console) shows timeouts for this test on Connext.

File: rclcpp_examples/test/test_executables_example.py.in [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+                                              expected_outputs):
+        output_lines = handler.stdout_data.getvalue().splitlines()
+    
+        assert rc == 0, \
+            "The launch file failed with exit code '" + str(rc) + "'. " \
+            'Maybe the client did not receive any messages?'
+    
+        assert output_lines == expected_output_lines, \
+            "Example output (%r) does not match expected output (%r)" % \
+            (output_lines, expected_output_lines)
```
Comment (dirk-thomas) [inline rclcpp_examples/test/test_executables_example.py.in]: Isn't relying on the first message (`I heard: [Hello World: 1]`) to arrive a race condition?

Comment (gerkey) [inline rclcpp_examples/test/test_executables_example.py.in]: Yep.  That's why the test is failing with Connext (where we lose the race).  It has so far worked every time on OpenSplice (because we win the race), but that's likely not guaranteed.

I'll fix that next.

**Comment by gerkey** [comment]
Ideas for broadening the output-matching scheme to allow tests like the talker / listener test (which don't always produce exactly the same output):

* Convert from templated Python program to Python library and for each test write a Python script that calls into that library, specifying things like how to match output.
* In addition to `.txt` files that are just matched, allow output files to be `.py` and in that case import them and call into functions by known names to match output.
* Allow regex or glob syntax in the `.txt` files.

**Comment by esteve** [comment]
I've changed the scripts so that if instead of a `.txt` file, a Python script is found in the path, it'll be loaded and run to check the output of the example. For the `listener` example, the script just checks that the last received line is ''I heard: [Hello World: 2]' and kill the process accordingly. The tests pass now.

**Comment by esteve** [comment]
OSX CI job:

http://ci.ros2.org/job/ros2_batch_ci_osx/474/

File: rclcpp_examples/test/listener.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
+import io
+from launch.output_handler import LineOutput
+
+
+class CheckerHandler(LineOutput):
+    EXPECTED_LAST_LINE = b'I heard: [Hello World: 2]'
```
Comment (dirk-thomas) [inline rclcpp_examples/test/listener.py]: Same potential race here.

Comment (esteve) [inline rclcpp_examples/test/listener.py]: So, should we let talker run for longer?

Comment (esteve) [inline rclcpp_examples/test/listener.py]: Or just check that the listener receives `b'I heard: [Hello World: %d]'`, and not worry about the number?

Comment (dirk-thomas) [inline rclcpp_examples/test/listener.py]: Checking for any number in that string will be the robust approach.

Comment (esteve) [inline rclcpp_examples/test/listener.py]: Done.

**Comment by dirk-thomas** [comment]
What cases do we expect for the other demos / example to test? If the Python checker script necessary for those or would a text line with one-line-per-regex be sufficient for all the cases we currently foresee?

**Comment by esteve** [comment]
Most of the other examples follow the same pattern, just match the output (regex or literal matching). The demos that use the camera may prove to be a bit trickier.

**Comment by dirk-thomas** [comment]
If multiple regex (or even a single one) is sufficient then I would recommend only reading that form the text file and avoid the dynamic loading of additional Python code.

**Comment by esteve** [comment]
This is still in progress, so I was still working on what would be the best approach. Rewriting the handler to always use a regex for matching was one of the options.

**Comment by esteve** [comment]
Not ready for review yet, but here are the CI jobs:

http://ci.ros2.org/job/ros2_batch_ci_linux/592/
http://ci.ros2.org/job/ros2_batch_ci_osx/479/
http://ci.ros2.org/job/ros2_batch_ci_windows/654/

**Comment by esteve** [comment]
Ready for review now. Tests pass, except for an unrelated cpplint warning (https://github.com/ros2/rosidl/pull/50)

**Comment by esteve** [comment]
I moved the string parsing code to Python because I don't find CMake very readable for that, but I'm unsure about it, so I left it as a separate commit so I can revert it. I squashed all the other commits.

File: rclcpp_examples/test/test_executables_example.py.in [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+            # Filter out stdout that comes from underlying DDS implementations
+            if line.startswith(b'pid'):
+                continue
+            if line.startswith(b'rc'):
+                continue
+            if line.startswith(b'RTI Data Distribution Service Evaluation ' +
+                               b'License issued to'):
+                continue
+            if line.startswith(b'Expires on'):
+                continue
```
Comment (dirk-thomas) [inline rclcpp_examples/test/test_executables_example.py.in]: For shorter code:

    filter_lines_starting_with = [...]
    if any([line.startswith(prefix) for prefix in filter_lines_starting_with]):
        continue

File: rclcpp_examples/test/test_executables_example.py.in [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+            self.stdout_data.write(line)
+            # Add back the newline that was stripped by splitlines()
+            self.stdout_data.write(b'\n')
```
Comment (dirk-thomas) [inline rclcpp_examples/test/test_executables_example.py.in]: For shorter code:

    self.stdout_data.write(line + b'\n')

File: rclcpp_examples/test/test_executables_example.py.in [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    def on_stderr_lines(self, lines):
+        self.stderr_data.write(lines)
+
+    def get_description(self):
+        return 'InMemoryHandler'
+
+    def check(self):
+        output_lines = self.stdout_data.getvalue().splitlines()
+        assert self.matched, \
+            "Example output (%r) does not match expected output (%r)" % \
```
Comment (dirk-thomas) [inline rclcpp_examples/test/test_executables_example.py.in]: Single quotes.

File: rclcpp_examples/test/test_executables_example.py.in [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        output_file = os.path.join(examples_expected_output_path, exe_name)
+        name = 'test_executable_' + str(i)
+        literal_file = output_file + '.txt'
+        regex_file = output_file + '.regex'
+        py_file = output_file + '.py'
+        handler = None
+        if os.path.isfile(literal_file):
+            with open(literal_file, 'rb') as f:
+                expected_output = f.read().splitlines()
+                handler = InMemoryHandler(name, ld, expected_output, False)
```
Comment (dirk-thomas) [inline rclcpp_examples/test/test_executables_example.py.in]: Doesn't need to be within the `with` block.

Same below.

File: rclcpp_examples/test/test_executables_example.py.in [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+        elif os.path.isfile(regex_file):
+            with open(regex_file, 'rb') as f:
+                expected_output = f.read().splitlines()
+                handler = InMemoryHandler(name, ld, expected_output, True)
+        elif os.path.isfile(py_file):
+            checker_module = SourceFileLoader(
+                "checker_module", py_file).load_module()
+            handler = checker_module.CheckerHandler(name, ld)
+        else:
+            assert handler, "Cannot find appropriate handler"
```
Comment (dirk-thomas) [inline rclcpp_examples/test/test_executables_example.py.in]: Single quotes.

**Comment by dirk-thomas** [comment]
Should the whole *handler* code / logic be moved somewhere else in order to allow reusing it for the next test?

**Comment by esteve** [comment]
@dirk-thomas sounds good, where should this code go? It might be reused for the demos tests, should it be a separate package?

**Comment by dirk-thomas** [comment]
May be a new package `launch_test` / `launch_testing` in the `lauch` repo?

**Comment by esteve** [comment]
@dirk-thomas makes sense, I'll create that.

**Comment by esteve** [comment]
Updated to use the the `launch_testing` package

**Comment by esteve** [comment]
I've reverted the string parsing code, so it now is in CMake again.

**Comment by dirk-thomas** [comment]
Currently the CMake generates Python code (at least a list of strings). I wouldn't do that in CMake as you have mentioned it is not very nice code. Why not create a CMake list within CMake and set pass that to Python and let Python do a simple split on semicolon to iterate over the list?

**Comment by esteve** [comment]
@dirk-thomas I've changed the code so that CMake no longer generates Python, just concatenates paths with executables and output logs, and the Python counterpart splits the strings.

**Comment by esteve** [comment]
CI:

http://ci.ros2.org/job/ros2_batch_ci_linux/594/
http://ci.ros2.org/job/ros2_batch_ci_osx/481/
http://ci.ros2.org/job/ros2_batch_ci_windows/658/

**Comment by esteve** [comment]
The CI jobs failed because the tests depend on https://github.com/ros2/launch/pull/13

File: rclcpp_examples/CMakeLists.txt [hunk 8]
```diff
@@ -101,23 +101,35 @@ if(AMENT_ENABLE_TESTING)
...
 
-  foreach(example_executable list_parameters_async list_parameters parameter_events_async parameter_events set_and_get_parameters_async set_and_get_parameters)
+  # Add each test case.  Multi-executable tests can be specified in
+  # colon-separated strings, like  exe1:exe2.
+  foreach(example_executable list_parameters_async list_parameters parameter_events_async parameter_events set_and_get_parameters_async set_and_get_parameters talker:listener)
+
+    set(RCLCPP_EXAMPLES_EXPECTED_OUTPUT "${CMAKE_CURRENT_SOURCE_DIR}/test/${example_executable}")
+    string(
+      REPLACE ":" ":${CMAKE_CURRENT_SOURCE_DIR}/test/"
+      RCLCPP_EXAMPLES_EXPECTED_OUTPUT ${RCLCPP_EXAMPLES_EXPECTED_OUTPUT})
```
Comment (dirk-thomas) [inline rclcpp_examples/CMakeLists.txt]: Instead of this custom logic with a colon separator and then applying the same prefix twice (L108 and L109-111) I would suggest using something more "native". Simply making each element a CMake list:

    foreach(example_executable list_parameters_async list_parameters parameter_events_async parameter_events set_and_get_parameters_async set_and_get_parameters "talker;listener")

And then apply the same operation on each item of the list. That would avoid duplication of logic and help to make the code more readable.

Same below.

Comment (esteve) [inline rclcpp_examples/CMakeLists.txt]: I've changed the CMake to use a list of lists and use a `foreach` loop instead of `string(REPLACE...)`. Please let me know if that's more readable. Thanks!

File: rclcpp_examples/CMakeLists.txt [hunk 9]
```diff
@@ -101,23 +101,36 @@ if(AMENT_ENABLE_TESTING)
...
   endforeach()
 
-  foreach(example_executable list_parameters_async list_parameters parameter_events_async parameter_events set_and_get_parameters_async set_and_get_parameters)
+  # Add each test case.  Multi-executable tests can be specified in
+  # colon-separated strings, like  exe1:exe2.
+  foreach(example_executable list_parameters_async list_parameters parameter_events_async parameter_events set_and_get_parameters_async set_and_get_parameters "talker;listener")
+
+    set(RCLCPP_EXAMPLES_EXPECTED_OUTPUT "")
+    foreach(executable ${example_executable})
+      set(RCLCPP_EXAMPLES_EXPECTED_OUTPUT "${RCLCPP_EXAMPLES_EXPECTED_OUTPUT} ${CMAKE_CURRENT_SOURCE_DIR}/test/${executable}")
```
Comment (dirk-thomas) [inline rclcpp_examples/CMakeLists.txt]: Why did you change it to space separators again? It simply doesn't work if you have spaces in the path.

After changing it back to a CMake list you might want to use `list(APPEND ...)` instead.

Comment (esteve) [inline rclcpp_examples/CMakeLists.txt]: I didn't understand what you meant before, please tell me exactly what the code should look like and I'll apply the changes.

Comment (dirk-thomas) [inline rclcpp_examples/CMakeLists.txt]: list(APPEND RCLCPP_EXAMPLES_EXPECTED_OUTPUT "${CMAKE_CURRENT_SOURCE_DIR}/test/${executable}")

Comment (esteve) [inline rclcpp_examples/CMakeLists.txt]: Updated. Thanks.

File: rclcpp_examples/test/test_executables_example.py.in [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+        elif os.path.isfile(regex_file):
+            with open(regex_file, 'rb') as f:
+                expected_output = f.read().splitlines()
+            handler = InMemoryHandler(name, ld, expected_output, True)
+        elif os.path.isfile(py_file):
+            checker_module = SourceFileLoader(
+                'checker_module', py_file).load_module()
+            handler = checker_module.CheckerHandler(name, ld)
+        else:
+            assert handler, 'Cannot find appropriate handler'
```
Comment (dirk-thomas) [inline rclcpp_examples/test/test_executables_example.py.in]: `handler` is always `None` here. So this could be:

    assert False, 'Cannot find appropriate handler for: %s' % output_file

File: rclcpp_examples/test/test_executables_example.py.in [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+    ld = LaunchDescriptor()
+
+    i = 0
+    for exe, output_file in zip('@RCLCPP_EXAMPLES_EXECUTABLE@'.split(';'),
+                                '@RCLCPP_EXAMPLES_EXPECTED_OUTPUT@'.split(';')):
+        name = 'test_executable_' + str(i)
+        literal_file = output_file + '.txt'
+        regex_file = output_file + '.regex'
+        py_file = output_file + '.py'
+        handler = None
```
Comment (dirk-thomas) [inline rclcpp_examples/test/test_executables_example.py.in]: Not necessary.

File: rclcpp_examples/test/test_executables_example.py.in [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+
+
+def test_executable():
+    output_handlers = []
+
+    ld = LaunchDescriptor()
+
+    i = 0
+    for exe, output_file in zip('@RCLCPP_EXAMPLES_EXECUTABLE@'.split(';'),
+                                '@RCLCPP_EXAMPLES_EXPECTED_OUTPUT@'.split(';')):
```
Comment (dirk-thomas) [inline rclcpp_examples/test/test_executables_example.py.in]: The generated code might get pretty long. The following would be easier to read.

    executables = '@RCLCPP_EXAMPLES_EXECUTABLE@'.split(';')
    output_files = '@RCLCPP_EXAMPLES_EXPECTED_OUTPUT@'.split(';'))
    for exe, output_file in zip(executables, output_files):                            :

File: rclcpp_examples/test/test_executables_example.py.in [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+def setup():
+    os.environ['OSPL_VERBOSITY'] = '8'  # 8 = OS_NONE
+
+
+def test_executable():
+    output_handlers = []
+
+    ld = LaunchDescriptor()
+
+    i = 0
```
Comment (dirk-thomas) [inline rclcpp_examples/test/test_executables_example.py.in]: You could use `enumerate` to avoid maintaining your own counter variable.

File: rclcpp_examples/test/test_executables_example.py.in (hunk 10 above)
Comment (esteve) [inline rclcpp_examples/test/test_executables_example.py.in]: Done.

File: rclcpp_examples/test/test_executables_example.py.in (hunk 11 above)
Comment (esteve) [inline rclcpp_examples/test/test_executables_example.py.in]: Done.

File: rclcpp_examples/test/test_executables_example.py.in (hunk 12 above)
Comment (esteve) [inline rclcpp_examples/test/test_executables_example.py.in]: Done.

File: rclcpp_examples/test/test_executables_example.py.in (hunk 13 above)
Comment (esteve) [inline rclcpp_examples/test/test_executables_example.py.in]: Done.

**Comment by esteve** [comment]
CI jobs:

http://ci.ros2.org/job/ros2_batch_ci_linux/597/
http://ci.ros2.org/job/ros2_batch_ci_osx/484/
http://ci.ros2.org/job/ros2_batch_ci_windows/665/

**Comment by esteve** [comment]
Tests passing after adding `launch` as an exec dependency for `launch_testing`:

http://ci.ros2.org/job/ros2_batch_ci_linux/615/
http://ci.ros2.org/job/ros2_batch_ci_osx/507/
http://ci.ros2.org/job/ros2_batch_ci_windows_opensplice/683/

**Comment by dirk-thomas** [comment]
lgtm

**Comment by esteve** [comment]
Forgot to link the CI jobs here:

http://ci.ros2.org/job/ros2_batch_ci_linux/617/
http://ci.ros2.org/job/ros2_batch_ci_osx/508/
http://ci.ros2.org/job/ros2_batch_ci_windows_opensplice/684/
